src/gt4py/next/iterator/transforms/fuse_maps.py

```diff
@@ -0,0 +1,128 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2023, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import dataclasses
+from typing import TypeGuard
+
+from gt4py.eve import NodeTranslator, traits
+from gt4py.eve.utils import UIDGenerator
+from gt4py.next.iterator import ir
+from gt4py.next.iterator.transforms import inline_lambdas
+
+
+def _is_map(node: ir.Node) -> TypeGuard[ir.FunCall]:
+    return (
+        isinstance(node, ir.FunCall)
+        and isinstance(node.fun, ir.FunCall)
+        and node.fun.fun == ir.SymRef(id="map_")
+    )
+
+
+def _is_reduce(node: ir.Node) -> TypeGuard[ir.FunCall]:
+    return (
+        isinstance(node, ir.FunCall)
+        and isinstance(node.fun, ir.FunCall)
+        and node.fun.fun == ir.SymRef(id="reduce")
+    )
+
+
+@dataclasses.dataclass(frozen=True)
+class FuseMaps(traits.VisitorWithSymbolTableTrait, NodeTranslator):
+    """
+    Fuses nested `map_`s.
+
+    Preconditions:
+      - `FunctionDefinitions` are inlined
+
+    Example:
+        map(λ(x, y)->f(x, y))(a, map(λ(z, w)->g(z, w))(b, c))
+    to
+        map(λ(a, b, c) → f(a, g(b, c)))(a, b, c)
+    """
+
+    uids: UIDGenerator = dataclasses.field(init=False, repr=False, default_factory=UIDGenerator)
+
+    def _as_lambda(self, fun: ir.SymRef | ir.Lambda, param_count: int) -> ir.Lambda:
+        if isinstance(fun, ir.Lambda):
+            return fun
+        params = [ir.Sym(id=self.uids.sequential_id(prefix="sym")) for _ in range(param_count)]
```

Comment:
```suggestion
        params = [ir.Sym(id=self.uids.sequential_id(prefix="_fuse_maps")) for _ in range(param_count)]
```
The current (fragile) approach is to use a unique prefix per pass and only construct the pass once. Just using `sym` looks dangerous the me as `fun` might capture a `sym` from outside. I would also add a precondition in the docstring above reflecting that.